chore: add docs about runtime upgrades (#139)
chainflip-io/chainflip-docs: pages/mainnet/validator-setup/managing-runtime-upgrades.mdx

---

import { Callout } from "@/components";

# The Problem

Periodically the Chainflip Labs team will need to perform a Runtime Upgrade to the node. You can read more about this [here](https://docs.substrate.io/maintain/runtime-upgrades/).
Unfortunately it is very difficult to make the `chainflip-engine` backwards compatible with older versions of the runtime.
Therefore, we require operators to run both the new and old version of the engine simultaneously.
This guide walks through doing this on VPS and Kubernetes.

<Callout type="info">This process is only required for Major or Minor releases (i.e. 1.x -> 1.y)</Callout>

## VPS

Upgrading on VPS is very trivial. We supply the new binaries as an entirely separate package. The following shows how you
would upgrade from `1.0.0` to `1.1.0`.

### Before the Runtime Upgrade

When we announce that a new version of the software is available you can already get ready by running the second version of the engine.

```shell
sudo apt update
sudo apt --only-upgrade install chainflip-node
sudo apt install chainflip-engine1.1
```

> You should see the following logs from the engine:

Reviewer: ```suggestion
You should see the following logs from the new engine:
```

```shell

> jouanalctl -f -u chainflip-engine1.1

Reviewer: ```suggestion
journalctl -f -u chainflip-engine1.1
```

{"timestamp":"2023-12-11T16:26:38.381549Z","level":"INFO","fields":{"message":"This version '1.1.0' is incompatible with the current release '1.0.0' at block: 0x4d66…fb38. WAITING for a compatible release version."},"target":"chainflip_engine::state_chain_observer::client"}
```

### After the Runtime Upgrade

> Check the logs of the new version of the `chainflip-engine1.1`. You should see that it started successfully.

Reviewer: ```suggestion
Check the logs of the new version of the `chainflip-engine1.1`. You should see that it is now processing blocks.
```

```shell
sudo systemctl disable --now chainflip-engine1.0.service
sudo apt purge chainflip-engine1.0
```

## Kubernetes
While we don't officially support running Validators on Kubernetes, we understand its appeal.

The main difficulty is having two containers that can both access the `data.db` directory. Sadly, most cloud providers
don't support `ReadWriteMany` on volumes. So the next best thing would be to run two containers in a single pod. The issue
here is that there is no easy way to route traffic from the `Service` to the correct pod at the moment the Runtime Upgrade
is pushed. It would require the operator to be available at the time of the upgrade which can be difficult considering we
are a global network.

The best solution we can offer is to build a custom image wherein both versions of the `chainflip-engine` can run simultaneously.
Docker (while also making it clear that this isn't best practise), provides several solutions on their [website](https://docs.docker.com/config/containers/multi-service_container/).

> Of the listed soltions, using [Supervisord](http://supervisord.org/) suits us the best.

Reviewer: ```suggestion
Of the listed solutions, using [Supervisord](http://supervisord.org/) suits us the best.
```

    rm -rf /var/lib/apt/lists/*

WORKDIR /etc/chainflip

RUN chmod +x /usr/local/bin/chainflip-engine \
    && useradd -m -u 1000 -U -s /bin/sh -d /flip flip \
    && chown -R 1000:1000 /etc/chainflip \
    && rm -rf /sbin /usr/sbin /usr/local/sbin

USER flip

COPY --from=chainflip-engine1-0-0 --chown=1000:1000 /usr/local/bin/chainflip-engine /usr/local/bin/chainflip-engine1.0
COPY --from=chainflip-engine1-1-0 --chown=1000:1000 /usr/local/bin/chainflip-engine /usr/local/bin/chainflip-engine1.1

COPY supervisord.conf .

ENTRYPOINT ["supervisord", "-c", "supervisord.conf"]
```

```toml
[supervisord]
nodaemon=true
logfile=/dev/null
logfile_maxbytes=0

[program:engine_1]
command=/usr/local/bin/chainflip-engine1.1
stdout_logfile=/dev/fd/1
stdout_logfile_maxbytes=0
redirect_stderr=true
startretries=10
startsecs=10
autorestart=true
priority=1

[program:engine_2]
command=/usr/local/bin/chainflip-engine1.0
stdout_logfile=/dev/fd/1
stdout_logfile_maxbytes=0
redirect_stderr=true
startretries=10
startsecs=10
autorestart=true
priority=2

[supervisorctl]
```

To build simply run:

```shell
docker buildx build -t my_repo/my_image:tag .
```

### Before the Runtime Upgrade

> Make sure you are running your custom image once the team has made an announcement.

Reviewer: Maybe mention that the new engine will register itself on-chain to signal its readiness for the upgrade.

> Make sure you are running your custom image once the team has made an announcement.

Author: ```suggestion
Make sure you are running your custom image once the team has made an announcement. The new engine will register itself on-chain thereby signalling its readiness for the upgrade.
```
